### PR TITLE
Dice cost

### PR DESCRIPTION
**Pull request by FANG-MING**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add an option to `DiceCost` which is used in K-Net, by setting `naive_dice` to `False` to
use it.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet/core/bbox/assigners/mask_hungarian_assigner.py [hunk 1]
```diff
@@ -70,15 +70,15 @@ def assign(self,
...
             'Only case when gt_bboxes_ignore is None is supported.'
-        num_gt, num_query = gt_labels.shape[0], cls_pred.shape[0]
+        num_gt, num_query = gt_labels.shape[0], mask_pred.shape[0]
 
         # 1. assign -1 by default
-        assigned_gt_inds = cls_pred.new_full((num_query, ),
+        assigned_gt_inds = mask_pred.new_full((num_query, ),
+                                              -1,
+                                              dtype=torch.long)
+        assigned_labels = mask_pred.new_full((num_query, ),
```
Comment (jbwang1997) [inline mmdet/core/bbox/assigners/mask_hungarian_assigner.py:79]: Is this change necessary?

File: mmdet/core/bbox/match_costs/match_cost.py [hunk 2]
```diff
@@ -244,12 +295,14 @@ class DiceCost:
         pred_act (bool, optional): Whether to apply sigmoid to mask_pred.
             Defaults to False.
         eps (float, optional): default 1e-12.
+        naive_dice (bool, optional): default True.
```
Comment (chhluo) [inline mmdet/core/bbox/match_costs/match_cost.py:298]: add description about difference between dice and naive_dice

File: mmdet/core/bbox/match_costs/match_cost.py [hunk 3]
```diff
@@ -7,6 +7,57 @@
 from .builder import MATCH_COST
 
 
+@MATCH_COST.register_module()
```
Comment (chhluo) [inline mmdet/core/bbox/match_costs/match_cost.py:10]: Is it same as `CrossEntropyLossCost`?

File: mmdet/core/bbox/assigners/mask_hungarian_assigner.py (hunk 1 above)
Comment (ZwwWayne) [reply mmdet/core/bbox/assigners/mask_hungarian_assigner.py:79]: K-Net sometimes passes cls_pred=None to this assigner.
We should leave comments here to indicate that why we use the shape of mask_pred

File: mmdet/core/bbox/match_costs/match_cost.py [hunk 4]
```diff
@@ -244,12 +244,18 @@ class DiceCost:
         pred_act (bool, optional): Whether to apply sigmoid to mask_pred.
             Defaults to False.
         eps (float, optional): default 1e-12.
+        naive_dice (bool, optional): If false, use the dice \
+                loss defined in the V-Net paper, otherwise, use the \
+                naive dice loss in which the power of the number in the
+                denominator is the first power instead of the second
```
Comment (ZwwWayne) [inline mmdet/core/bbox/match_costs/match_cost.py:250]: in which the denominator is calculated by the first power instead of the second power of mask values.

File: mmdet/core/bbox/match_costs/match_cost.py [hunk 5]
```diff
@@ -265,7 +271,12 @@ def binary_mask_dice_loss(self, mask_preds, gt_masks):
         mask_preds = mask_preds.flatten(1)
         gt_masks = gt_masks.flatten(1).float()
         numerator = 2 * torch.einsum('nc,mc->nm', mask_preds, gt_masks)
-        denominator = mask_preds.sum(-1)[:, None] + gt_masks.sum(-1)[None, :]
+        if self.naive_dice:
+            denominator = mask_preds.sum(-1)[:, None] + \
+                gt_masks.sum(-1)[None, :]
+        else:
+            denominator = torch.sum(mask_preds * mask_preds, 1)[:, None] + \
```
Comment (ZwwWayne) [inline mmdet/core/bbox/match_costs/match_cost.py:278]: torch.sum(mask_preds * mask_preds, 1) -> mask_preds.pow(2).sum(1), also update that for gt_masks

File: mmdet/core/bbox/match_costs/match_cost.py [hunk 6]
```diff
@@ -244,12 +244,17 @@ class DiceCost:
         pred_act (bool, optional): Whether to apply sigmoid to mask_pred.
             Defaults to False.
         eps (float, optional): default 1e-12.
+        naive_dice (bool, optional): If True, use the dice \
```
Comment (ZwwWayne) [inline mmdet/core/bbox/match_costs/match_cost.py:247]: If True, use the naive dice loss in which the power of the number in the denominator is the first power.
If Flase, use the second power that is adopted by K-Net and SOLO. Defaults to True.

File: tests/test_utils/test_assigner.py [hunk 7]
```diff
@@ -628,3 +628,15 @@ def test_mask_hungarian_match_assigner():
...
+    # test with dicecost used in K-Net
+    assigner_cfg = dict(
+        cls_cost=dict(type='ClassificationCost', weight=0.0),
+        dice_cost=dict(
+            type='DiceCost',
+            weight=0.0,
+            pred_act=True,
+            eps=1.0,
+            naive_dice=False))
+    with pytest.raises(AssertionError):
```
Comment (ZwwWayne) [inline tests/test_utils/test_assigner.py:641]: We should not only test AssertionError, we should also test the case when it can run normally.

File: tests/test_utils/test_assigner.py [hunk 8]
```diff
@@ -628,3 +628,15 @@ def test_mask_hungarian_match_assigner():
         dice_cost=dict(type='DiceCost', weight=0.0, pred_act=True, eps=1.0))
     with pytest.raises(AssertionError):
         self = MaskHungarianAssigner(**assigner_cfg)
+
+    # test with dicecost used in K-Net
+    assigner_cfg = dict(
```
Comment (chhluo) [inline tests/test_utils/test_assigner.py:633]: Please refer to https://github.com/open-mmlab/mmdetection/blob/3e2693151add9b5d6db99b944da020cba837266b/tests/test_utils/test_assigner.py#L598.
Put test for `naive_dice=True, naive_dice=False` together

**Review by chhluo — APPROVED** [review]

**Review by chhluo — APPROVED** [review]

**Comment by OpenMMLab-gd** [comment]
Dear FANG-MING,
First of all, we want to express our gratitude for your significant PR in the MMDetection project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.
If you are Chinese or have WeChat，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. We look forward to seeing you there! Join us ：https://discord.gg/raweFPmdzG
Thank you again for your contribution❤
Best regards！ @FANG-MING
